From 388684b798b2d93d6c6ef3785eda9cd608a341a9 Mon Sep 17 00:00:00 2001
From: Ivo Ivanov <logos128@gmail.com>
Date: Thu, 18 Aug 2022 15:48:08 +0300
Subject: [PATCH] ntoskrnl.exe: Prevent blocking if timer is reset in DPC.

Fixes NaturalPoint's TrackIR5 app hanging indefinitely on exit,
while waiting synchronously on an IRP to complete. The blocking
happens in the app's npusbio_x64 driver, while it tries to reset
a timer in its DPC.
Probably fixes other drivers/apps in such situations.
---
 dlls/ntoskrnl.exe/sync.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/dlls/ntoskrnl.exe/sync.c b/dlls/ntoskrnl.exe/sync.c
index 11111111111..11111111111 100644
--- a/dlls/ntoskrnl.exe/sync.c
+++ b/dlls/ntoskrnl.exe/sync.c
@@ -493,10 +493,14 @@ BOOLEAN WINAPI KeSetTimerEx( KTIMER *timer, LARGE_INTEGER duetime, LONG period,
 
     EnterCriticalSection( &sync_cs );
 
-    if ((ret = timer->Header.Inserted))
-        KeCancelTimer(timer);
-
+    ret = timer->Header.Inserted;
     timer->Header.Inserted = TRUE;
+    timer->Header.SignalState = FALSE;
+    if (timer->Header.WaitListHead.Blink && !*((ULONG_PTR *)&timer->Header.WaitListHead.Flink))
+    {
+        CloseHandle(timer->Header.WaitListHead.Blink);
+        timer->Header.WaitListHead.Blink = NULL;
+    }
 
     if (!timer->TimerListEntry.Blink)
         timer->TimerListEntry.Blink = (void *)CreateThreadpoolTimer(ke_timer_complete_proc, timer, NULL);
-- 
GitLab
